src/allocation_engine/interval_tree.rs

```diff
@@ -185,8 +185,173 @@ struct InnerNode {
     right: Option<Box<InnerNode>>,
     /// Cached height of the node.
     height: u64,
-    /// Cached maximum valued covered by this node.
-    max_key: u64,
+}
+
+impl InnerNode {
+    /// Creates a new InnerNode object.
+    #[allow(dead_code)]
+    fn new(key: Range, node_state: NodeState) -> Self {
+        InnerNode {
+            key,
+            node_state,
+            left: None,
+            right: None,
+            height: 1,
+        }
+    }
+
+    /// Returns a readonly reference to the node associated with the `key` or
+    /// None if the searched key does not exist in the tree.
+    #[allow(dead_code)]
+    fn search(&self, key: &Range) -> Option<&InnerNode> {
+        match self.key.cmp(key) {
+            Ordering::Equal => Some(self),
+            Ordering::Less => self.right.as_ref().and_then(|node| node.search(key)),
+            Ordering::Greater => self.left.as_ref().and_then(|node| node.search(key)),
+        }
+    }
+
+    /// Returns a readonly reference to the node associated with the `key` or
+    /// None if there is no Node representing an interval that covers the
+    /// searched key. For a key [a, b], this method will return a node with
+    /// a key [c, d] such that c <= a and b <= d.
+    #[allow(dead_code)]
+    fn search_superset(&self, key: &Range) -> Option<&InnerNode> {
+        if self.key.contains(key) {
+            Some(self)
+        } else if key.end < self.key.start && self.left.is_some() {
+            // Safe to unwrap() because we have just checked it.
+            self.left.as_ref().unwrap().search_superset(key)
+        } else if key.start > self.key.end && self.right.is_some() {
+            // Safe to unwrap() because we have just checked it.
+            self.right.as_ref().unwrap().search_superset(key)
+        } else {
+            None
+        }
+    }
+
+    /// Rotate the tree such that height difference between subtrees
+    /// is not greater than abs(1).
+    #[allow(dead_code)]
+    fn rotate(self) -> Box<Self> {
+        let l = height(&self.left);
+        let r = height(&self.right);
+
+        match (l as i32) - (r as i32) {
+            1 | 0 | -1 => Box::new(self),
+            2 => self.rotate_left_successor(),
+            -2 => self.rotate_right_successor(),
+            _ => unreachable!(),
+        }
+    }
+
+    /// Perform a single left rotation on this node.
+    #[allow(dead_code)]
+    fn rotate_left(mut self) -> Box<Self> {
+        let mut new_root = self.right.take().expect("Node is broken");
+        self.right = new_root.left.take();
+        self.update_cached_info();
+        new_root.left = Some(Box::new(self));
```

Comment:
I'm a bit concerned with the amount of `Box::new()` calls in the `rotate*` methods. Each call is a memory allocation and is relatively expensive. Conceptually, you need to allocate memory only when instantiating a new node. The rotations should be performed by adjusting pointers, there should be no need for more allocations.

Now, I know that Rust is being, well, Rust, and simple things are usually difficult :) But could you please spend some time and investigate if it's possible to perform the rotations without allocations.